syft/pkg/cataloger/internal/dependency/resolver.go

```diff
@@ -0,0 +1,81 @@
+package dependency
+
+import (
+	"sort"
+
+	"github.com/scylladb/go-set/strset"
+
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/pkg/cataloger/generic"
+)
+
+// Prosumer is a producer and consumer, in this context, for packages that provide resources and require resources.
+type Prosumer interface {
+	Provides(pkg.Package) []string
+	Requires(pkg.Package) []string
+}
+
+type RelationshipResolver struct {
+	prosumer Prosumer
+}
+
+func NewRelationshipResolver(p Prosumer) RelationshipResolver {
+	return RelationshipResolver{
+		prosumer: p,
+	}
+}
+
+func Processor(p Prosumer) generic.Processor {
+	return func(pkgs []pkg.Package, rels []artifact.Relationship, err error) ([]pkg.Package, []artifact.Relationship, error) {
+		rels = append(rels, NewRelationshipResolver(p).Resolve(pkgs)...)
+		return pkgs, rels, err
+	}
+}
+
+// Resolve will create relationships between packages based on the "Depends" and "Provides" specifications from the given packages.
+func (r RelationshipResolver) Resolve(pkgs []pkg.Package) (relationships []artifact.Relationship) {
+	lookup := make(map[string][]pkg.Package)
+
+	for _, p := range pkgs {
+		for _, key := range deduplicate(r.prosumer.Provides(p)) {
+			lookup[key] = append(lookup[key], p)
+		}
+	}
+
+	seen := strset.New()
+	for _, p := range pkgs {
+		for _, requirement := range deduplicate(r.prosumer.Requires(p)) {
+			for _, depPkg := range lookup[requirement] {
+				// prevent creating duplicate relationships
+				pairKey := string(depPkg.ID()) + "-" + string(p.ID())
+				if seen.Has(pairKey) {
+					continue
+				}
+
+				relationships = append(relationships,
+					artifact.Relationship{
+						From: depPkg,
+						To:   p,
+						Type: artifact.DependencyOfRelationship,
+					},
+				)
+
+				seen.Add(pairKey)
+			}
+		}
+	}
+	return relationships
+}
+
+func deduplicate(ss []string) []string {
+	set := strset.New(ss...)
+	// note: this must be a stable function
```

Comment:
I don't understand this comment. If I have a deduplicated list of strings, then it doesn't matter whether it's sorted in a stable or unstable way, since there are no duplicate keys to break ties between. Or does "stable" in the comment not mean stable sorting? What am I missing here?

Reply:
there is no guarantee that `strset.List` will yield the same ordering on subsequent invocations. In this case we're interested in multiple invocations of syft... we want them to produce the same SBOM, thus be resistant to unstable ordering.

Reply:
I think I see what you're saying. You're saying, "The return value of `deduplcate` must be stable across multiple runs of Syft, but `set.List()` does not return a deterministic ordering. Therefore, sort the return value."

I think I was confused because the comment `// note; this must be a stable function` is right before the invocation of a function that is _not_ stable.

This is a tiny nit, but I think it would be clearer if the comment were moved to be above the call to `strings.Sort`, and said something like, "ensure order is stable across multiple invocations of Syft," or something like that.